### PR TITLE
Fixes #237 - support multi-dimensional arrays

### LINKED ISSUE #237: Support multi-dimensional arrays when writing json (example geojson polygon)
The current json-sink is not able to export Geojson correctly (probably because the cosmos source used Newtonsoft and writes using STJ).

Example json in cosmos
```json
{
  "id": "1",
  "name": "Example",
  "point": {
    "type": "Point",
    "coordinates": [
      5.347494076316281,
      52.033503157065155
    ]
  },
  "polygon": {
    "type": "Polygon",
    "coordinates": [
      [
        [
          5.347432321215393,
          52.03355800306437
        ],
        [
          5.347432321215393,
          52.03343276598602
        ],
        [
          5.347605671445933,
          52.03343276598602
        ],
        [
          5.347605671445933,
          52.03355800306437
        ],
        [
          5.347432321215393,
          52.03355800306437
        ]
      ]
    ]
  },
  "line": {
    "type": "LineString",
    "coordinates": [
      [
        5.3474399338950604,
        52.03355740411766
      ],
      [
        5.347590198744001,
        52.033439655450024
      ]
    ]
  }
}
```
 
Export settings:
```json
{
    "Source": "cosmos-nosql",
    "SourceSettings": {
        "ConnectionString": "--- replace ---",
        "Database": "--- replace ---",
        "Container": "--- replace ---"
    },
    "Sink": "JSON",
    "SinkSettings": {
        "FilePath": "D:\\export\\example.json",
        "IncludeNullFields": true,
        "Indented": true
    }
}

```

Result:
```json
[
  {
    "id": "1",
    "name": "Example",
    "point": {
      "type": "Point",
      "coordinates": [
        5.347494076316281,
        52.033503157065155
      ]
    },
    "polygon": {
      "type": "Polygon",
      "coordinates": [
        "System.Collections.Generic.List`1[System.Object]"
      ]
    },
    "line": {
      "type": "LineString",
      "coordinates": [
        "System.Collections.Generic.List`1[System.Object]",
        "System.Collections.Generic.List`1[System.Object]"
      ]
    }
  }
]
```

The export of Geojson Point is possible because the coordinates is a simple number array, but for Polygon (and others) the coordinates is a multi-dimensional array.

### PR DESCRIPTION
**Pull request by mlankamp**
Extended the `WriteFieldValue` function to support multi-dimensional arrays 

Fixes #237

### CODE REVIEW

**Comment by mlankamp** [comment]
@philnach & @markjbrown How do I request a review?

**Comment by philnach** [comment]
@mlankamp, thanks for submitting the PR and reaching out. you've done the right thing by opening a PR.  You could in the Reviewers panel on the PR request specific people [Requesting reviews from collaborators and organization members](https://docs.github.com/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review#requesting-reviews-from-collaborators-and-organization-members) to review your code.

With that said, I'll take a look.

**Review by philnach — CHANGES_REQUESTED** [review]
Could you add tests for this scenario to ensure the behavior works? Tests would go into this file: https://github.com/AzureCosmosDB/data-migration-desktop-tool/blob/main/Interfaces/Cosmos.DataTransfer.Common.UnitTests/DataItemJsonConverterTests.cs

**Comment by mlankamp** [comment]
Of course! Will add them tomorrow

File: Interfaces/Cosmos.DataTransfer.Common/DataItemJsonConverter.cs [hunk 1]
```diff
@@ -125,6 +133,10 @@ internal static void WriteFieldValue(Utf8JsonWriter writer, string fieldName, ob
                         var arrayDictItem = new DictionaryDataItem(arrayDict);
                         WriteDataItem(writer, arrayDictItem, includeNullFields);
                     }
+                    else if (arrayItem is not string && arrayItem is IEnumerable childs)
+                    {
+                        WriteFieldValue(writer, "", childs, includeNullFields);
```
Comment (philnach) [inline Interfaces/Cosmos.DataTransfer.Common/DataItemJsonConverter.cs:138]: Thanks for the PR — adding support for nested (multi-dimensional) arrays makes sense.

That said, I’m a bit concerned about the current approach of handling nested arrays by calling:

```csharp
WriteFieldValue(writer, "", childs, includeNullFields);
```

relies on:

```csharp
if (!string.IsNullOrWhiteSpace(propertyName.Value)) { writer.WriteStartArray(propertyName); }
else { writer.WriteStartArray(); }
```

Could we refactor nested-array writing so it doesn’t depend on passing fieldName == "" and the propertyName.Value check?

For example:

introduce a dedicated helper that writes an array value (no property name), e.g. `WriteArrayValue(Utf8JsonWriter writer, IEnumerable items, ...)`
use that helper for nested arrays so it always calls `writer.WriteStartArray() / writer.WriteEndArray()`

(Optional)
The nested-array detection `(arrayItem is not string && arrayItem is IEnumerable)` is very broad and may accidentally treat things like **byte[]** as arrays of numbers. It might be worth tightening this to the specific collection types we expect to see here.

**Comment by thucnguyen77** [comment]
Hi,

By any chance this PR can be merged soon?
I am working with GeoJSON and facing the same issue when trying to migrate data between Cosmos DB accounts.

Thanks,
Thuc

**Comment by philnach** [comment]
Hi @mlankamp — thanks again for the original fix and your patience. I've opened #243 which builds on your work and addresses the two review concerns:

1. **Avoids the sentinel `fieldName == """"` recursion** by splitting array writes into dedicated `WriteArrayItems` / `WriteArrayItemValue` helpers, so legitimate empty-string and whitespace property names are preserved.
2. **Adds a `MaxJsonDepth` guard** so deeply nested or recursive input throws a clean `InvalidOperationException` instead of stack-overflowing the host process — the failure mode I was worried about based on past Newtonsoft.Json bugs.

I also added a comprehensive test suite (17 new tests) covering the GeoJSON scenarios from #237, edge cases with nulls / mixed types / dictionaries inside nested arrays, regression guards for property-name handling, and at/over-max-depth tests for both arrays and objects.

Credit goes to you — I added a `Co-authored-by:` trailer on the commit. Closing this PR in favor of #243. Sorry for the long round-trip!

cc @thucnguyen77 — this should also fix the GeoJSON scenario you mentioned.
